bootstrap/tracing.go

```diff
@@ -0,0 +1,155 @@
+package bootstrap
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strconv"
+
+	"github.com/buildkite/agent/v3/agent"
+	"github.com/buildkite/agent/v3/env"
+	"github.com/buildkite/agent/v3/tracetools"
+	"github.com/opentracing/opentracing-go"
+	"golang.org/x/exp/slices"
+	ddext "gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/opentracer"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+)
+
+// stopper lets us abstract the tracer wrap up code so we can plug in different tracing
+// library implementations that are opentracing compatible. Opentracing itself
+// doesn't have a Stop function on its Tracer interface.
+type stopper func()
+
+func noopStopper() {}
+
+func (b *Bootstrap) startTracing(ctx context.Context) (tracetools.Span, context.Context, stopper) {
+	switch b.Config.TracingBackend {
+	case tracetools.BackendDatadog:
+		// Newer versions of the tracing libs print out diagnostic info which spams the
+		// Buildkite agent logs. Disable it by default unless it's been explicitly set.
+		if _, has := os.LookupEnv("DD_TRACE_STARTUP_LOGS"); !has {
+			os.Setenv("DD_TRACE_STARTUP_LOGS", "false")
+		}
+
+		return b.startTracingDatadog(ctx)
+
+	case tracetools.BackendNone:
+		return &tracetools.NoopSpan{}, ctx, noopStopper
+
+	default:
+		b.shell.Commentf("An invalid tracing backend was provided: %q. Tracing will not occur.", b.Config.TracingBackend)
+		b.Config.TracingBackend = "" // Ensure that we don't do any tracing after this, some of the stuff in tracetools uses the bootstrap's tracking backend
+		return &tracetools.NoopSpan{}, ctx, noopStopper
+	}
+}
+
+func (b *Bootstrap) getLabel() string {
+	label, ok := b.shell.Env.Get("BUILDKITE_LABEL")
+	if !ok {
+		label = "job"
+	}
+
+	return label
+}
+
+// startTracingDatadog sets up tracing based on the config values. It uses opentracing as an
+// abstraction so the agent can support multiple libraries if needbe.
+func (b *Bootstrap) startTracingDatadog(ctx context.Context) (tracetools.Span, context.Context, stopper) {
+	resourceName := b.OrganizationSlug + "/" + b.PipelineSlug + "/" + b.getLabel()
+	opts := []tracer.StartOption{
+		tracer.WithServiceName("buildkite_agent"),
+		tracer.WithSampler(tracer.NewAllSampler()),
+		tracer.WithAnalytics(true),
+		tracer.WithGlobalTag(ddext.ResourceName, resourceName),
```

Comment:
This used to be set on the root span, not as a global tag. I think that means the resource tag will be added to all spans, and thus count every span as a separate job run event. That's probably bad.

Reply:
that's a good catch, i didn't think about that at all. will change.